### ISSUE TITLE
Azure Blob Storage - find right archive structure, use .tar.gz (ratarmount) instead of .zip

### ISSUE THREAD

**Issue by percyliang** [codalab/codalab-worksheets#3112]
Currently, uploads and downloads are tar.gz, but in Azure blob store, they will be zip files.

**Comment by percyliang** [comment]
Need zip for random access.  Also look into bzip.

**Comment by epicfaace** [comment]
See https://blastedbio.blogspot.com/2013/04/random-access-to-blocked-xz-format-bxzf.html for more info about xz files.

**Comment by epicfaace** [comment]
Another option, which might actually be optimal because it keeps the extension .tar.gz: BGZF (https://blastedbio.blogspot.com/2011/11/bgzf-blocked-bigger-better-gzip.html)

A library for Python that works well with BGZF: https://biopython.org/docs/1.75/api/Bio.bgzf.html

**Comment by epicfaace** [comment]
@percyliang I've added a short writeup of what we can use for the archive structure at the top of this design doc: https://docs.google.com/document/d/1hfpfXBvoooIV6_cerKhnDu7X7t9o5hzGD7rtlTlDK_s/edit#

**Comment by epicfaace** [comment]
**Proposed solution:** Use .tar.gz files, but compress the tar files using ratarmount.
- ratarmount is a library that creates and stores an index to a .tar.gz file in an index.sqlite file.
- Both the index.sqlite and contents.tar.gz files will be stores on Blob Storage.
- To perform any operation (such as listing contents of a directory, downloading a file, etc.), the server will first download the index.sqlite file, read it, then read contents.tar.gz.

Discussion on ratarmount's repo about API changes required to do this: https://github.com/mxmlnkn/ratarmount/issues/49. I've implemented these functions in #3237, and made small changes to ratarmount (which I plan to get merged into ratarmount upstream).

To get this to work, the indexed_gzip library needs to be modified to allow the usage of file-like objects (essentially, the file-like objects that wrap Azure Blob Storage files). I modified the library in this PR (https://github.com/pauldmccarthy/indexed_gzip/pull/55). For now, I'll publish a forked version of the library that we can use in the initial implementation, but we'll switch back to the regular indexed_gzip library once my PR is merged.
